File: go/BSTIterator/BSTIterator.go
package main

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

type BSTIterator struct {
	Array []int
	CurrIdx int
}

func Constructor(root *TreeNode) *BSTIterator {
	array := make([]int, 0)

	// Time -> 0(N) Space -> 0(N)
	func InOrder(root *TreeNode, array *[]int) {
		if root == nil {
			return 
		}

		InOrder(root.Left, array)
		*array = append(*array, root.Val)
		InOrder(root.Right, array)
	}

	InOrder(root, &array)

	return &BSTIterator{Array: array, CurrIdx: -1}
}

// Time -> 0(1) Space -> 0(1)
func (this *BSTIterator) Next() int {
	this.CurrIdx++
	return this.Array[this.CurrIdx-1]
}

// Time -> 0(1) Space -> 0(1)
func (this *BSTIterator) HasNext() bool {
	return !len(this.Array) == this.CurrIdx
}

type BSTIteratorI struct {
	Array []*TreeNode
}

func ConstructorI(root *TreeNode) *BSTIteratorI {
	iterator := &BSTIteratorI{Array: []*TreeNode}
	iterator.LeftMost(root)
	return iterator
}

// Time -> 0(h) Space -> 0(h)
func (this *BSTIteratorI) LeftMost(root *TreeNode) {
	for root != nil {
		this.Array = append(this.Array, root.Val)
		root = root.Left
	}
}

// Time -> 0(h) Space -> 0(h) on avg 0(1)
func (this *BSTIteratorI) Next() int {
	curr := this.Array[len(this.Array)-1]
	this.Array = this[:len(this.Array)-1]
	this.LeftMost(curr.Right)
	return curr.Val
}

// Time -> 0(h) Space -> 0(1)
func (this *BSTIteratorI) HasNext() bool {
	return len(this.Array) > 0
}

File: python/bstIterator.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class BSTIterator:
    def __init__(self, root: TreeNode | None):
        self.array = self.get_inorder(root)
        self.curr_idx = 0

    # Time -> 0(N) Space -> 0(N)
    def get_inorder(self, root: TreeNode | None, array: list = []) -> list:
        if root is None:
            return []
        
        self.get_inorder(root.left, array)
        array.append(root.val)
        self.get_inorder(root.right, array)

        return array
    
    # Time -> 0(1) Space -> 0(1)
    def next(self) -> int:
        self.curr_idx += 1
        return self.array[self.curr_idx-1]
    
    # Time -> 0(1) Space -> 0(1)
    def has_next(self) -> bool:
        return not self.curr_idx == len(self.array)


class BSTIteratorI:
    def __init__(self, root: TreeNode | None):
        self.array = []
        self.leftmost(root)
        
    # Time -> 0(h) Space -> 0(h)
    def leftmost(self, root: TreeNode | None):
        while root is not None:
            self.array.append(root)
            root = root.left

    # Time -> 0(h) Space -> 0(h) on avg 0(1)
    def next(self) -> int:
        curr = self.array.pop()
        self.leftmost(curr.right)
        return curr.val
    
    # Time -> 0(h) Space -> 0(1)
    def has_next(self) -> bool:
        return len(self.array) > 0